### PR TITLE
Beepsky refactor and Bot code improvement

### PR DESCRIPTION
**Pull request by JohnFulpWillard**
## About The Pull Request

Refactors declare_arrests, idcheck, weaponscheck, check_records, arrest_type into flags
Makes Armsky a subtype of Beepsky rather than being varedited in every map
Re-organizes Robot defines to make up for this
Generally improves the code with early returns
Splits ED-209s into their own bot type rather than have a complete var decided to just saying if its a ED-209 or not
Adds a Sergeant-at-armsky in the Delta armory.
Merges var/baton_type and var/weapon, because I don't see why there were 2 separate vars for the same thing.

## Why It's Good For The Game

Bot code is by far the worst I've seen all year, I even had to make a whole new PR (https://github.com/tgstation/tgstation/pull/62223) just to ready myself for what this was gonna be. I plan on continuing this for even more bots and even the parent bot in general, but I'd like to stop here for now just to see if anyone's interested in me actually doing this.

## Changelog

:cl:
refactor: Bot emagging has been refactored, and as such, their cover can be closed if forced open by emag if the bot itself hasn't been emagged.
fix: Sergeant-At-Armsky can now be found in Delta's Armory!
/:cl:

The rest seems kinda not fit for changelog, which makes me wonder why 'Refactor' and 'Code Improvement' are still labels nowadays...

### CODE REVIEW

**Comment by TheBonded** [comment]
does this have any effect in game? I figure improving the code'd inadvertently make them faster/slightly better due to the code handling them better. not that this is a balance change, I'm just wondering if improving the code of secbots actually improves the way the game runs them, and therefore makes their AI slightly stronger

**Comment by JohnFulpWillard** [comment]
> does this have any effect in game? I figure improving the code'd inadvertently make them faster/slightly better due to the code handling them better. not that this is a balance change, I'm just wondering if improving the code of secbots actually improves the way the game runs them, and therefore makes their AI slightly stronger

I’m limited to testing testing this on a one man private server so I have no idea.

**Comment by SuperNovaa41** [comment]
> The rest seems kinda not fit for changelog, which makes me wonder why 'Refactor' and 'Code Improvement' are still labels nowadays

That sweet sweet GBP. Also because the players dont matter

**Review by Maurukas — APPROVED** [review]
Mapping looks fine as a simple replacement.

File: code/__DEFINES/robots.dm [hunk 1]
```diff
@@ -7,6 +7,49 @@
 #define LAW_ION "ion"
 #define LAW_HACKED "hacked"
 
+//AI notification defines
+///Alert when a new Cyborg is created.
+#define NEW_BORG 1
```
Comment (Mothblocks) [inline code/__DEFINES/robots.dm:12]: Can we please prefix all of these with something, some of them like `RENAME` and `DISCONNECT` are super generic

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 2]
```diff
@@ -90,32 +89,34 @@
...
-		var/list/mob/living/carbon/targets = new
-		for(var/mob/living/carbon/C in view(12,src))
-			if(C.stat==DEAD)
-				continue
-			targets += C
+	if(severity <= 1)
+		return
+	new /obj/effect/temp_visual/emp(loc)
+	var/list/mob/living/carbon/targets = new
+	for(var/mob/living/carbon/C in view(12,src))
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:96]: Better variable name

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 3]
```diff
@@ -90,32 +89,34 @@
...
-		new /obj/effect/temp_visual/emp(loc)
-		var/list/mob/living/carbon/targets = new
-		for(var/mob/living/carbon/C in view(12,src))
-			if(C.stat==DEAD)
-				continue
-			targets += C
+	if(severity <= 1)
+		return
+	new /obj/effect/temp_visual/emp(loc)
+	var/list/mob/living/carbon/targets = new
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:95]:
```suggestion
	var/list/mob/living/carbon/targets = list()
```

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 4]
```diff
@@ -90,32 +89,34 @@
...
+		if(C.stat == DEAD)
+			continue
+		targets += C
+	if(!targets.len)
+		return
+	if(prob(50))
+		var/mob/toshoot = pick(targets)
+		if(toshoot)
+			targets -= toshoot
+			if(prob(50) && emagged < 2)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:106]: `emagged` being set to numbers like 2 and whatnot is dumb, within scope to fix all that to *at least* defines?

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 5]
```diff
@@ -23,34 +23,44 @@
...
-	idcheck = FALSE
-	weaponscheck = FALSE
 	auto_patrol = TRUE
 	commissioned = TRUE
 
+/mob/living/simple_animal/bot/secbot/beepsky/armsky
+	name = "Sergeant-At-Armsky"
+	health = 45
+	auto_patrol = FALSE
+	security_mode_flags = SECBOT_DECLARE_ARRESTS | SECBOT_CHECK_IDS| SECBOT_CHECK_RECORDS
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:62]:
```suggestion
	security_mode_flags = SECBOT_DECLARE_ARRESTS | SECBOT_CHECK_IDS | SECBOT_CHECK_RECORDS
```

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 6]
```diff
@@ -252,17 +260,16 @@ Auto Patrol: []"},
...
-				stun_attack(A, TRUE)
-			else
-				stun_attack(A)
-		else if(C.canBeHandcuffed() && !C.handcuffed)
-			cuff(A)
-	else
-		..()
+	if(!iscarbon(A))
+		return ..()
+	var/mob/living/carbon/C = A
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:265]: Better variable name

File: code/modules/mob/living/simple_animal/bot/secbot.dm [hunk 7]
```diff
@@ -520,32 +518,31 @@ Auto Patrol: []"},
...
-				say("[market_verb] NAP Violation: No ID card found.")
-				nap_violation(target)
-				return FALSE
-	return TRUE
+	if(!target)
+		return TRUE
+	if(!ishuman(target))
+		return TRUE
+	var/mob/living/carbon/human/human_target = target
+	var/obj/item/card/id/target_id = human_target.get_idcard(TRUE)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/secbot.dm:526]: Use named argument here

File: code/modules/spells/spell_types/conjure.dm [hunk 8]
```diff
@@ -55,7 +55,14 @@
 	summon_type = list(/mob/living/simple_animal/bot/secbot/ed209)
 	summon_amt = 10
 	range = 3
-	newVars = list("emagged" = 2, "remote_disabled" = 1,"shoot_sound" = 'sound/weapons/laser.ogg',"projectile" = /obj/projectile/beam/laser, "declare_arrests" = 0,"name" = "Wizard's Justicebot")
+	new_vars = list(
+		"emagged" = 2,
+		"remote_disabled" = 1,
+		"shoot_sound" = 'sound/weapons/laser.ogg',
+		"projectile" = /obj/projectile/beam/laser,
+		security_mode_flags = ~(SECBOT_DECLARE_ARRESTS),
```
Comment (Mothblocks) [inline code/modules/spells/spell_types/conjure.dm:63]:
```suggestion
		"security_mode_flags" = ~(SECBOT_DECLARE_ARRESTS),
```

Be consistent

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm (hunk 4 above)
Comment (JohnFulpWillard) [reply code/modules/mob/living/simple_animal/bot/ed209bot.dm]: I removed emagged going up to 2 and its just TRUE/FALSE now. I can move it to a separate PR if wanted but I didn't see reason to keep it around if even Cyborg covers can be closed once emagged open.

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 9]
```diff
@@ -40,32 +40,32 @@
 /mob/living/simple_animal/bot/secbot/ed209/handle_automated_action()
 	var/judgement_criteria = judgement_criteria()
 	var/list/targets = list()
-	for(var/mob/living/carbon/C in view(7,src)) //Let's find us a target
+	for(var/mob/living/carbon/nearby_carbons in view(7, src)) //Let's find us a target
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:43]: `nearby_carbon`, it's only one.

File: code/modules/mob/living/simple_animal/bot/ed209bot.dm [hunk 10]
```diff
@@ -40,32 +40,32 @@
...
 		var/threatlevel = 0
-		if(C.incapacitated())
+		if(nearby_carbons.incapacitated())
 			continue
-		threatlevel = C.assess_threat(judgement_criteria, weaponcheck=CALLBACK(src, .proc/check_for_weapons))
-		//speak(C.real_name + text(": threat: []", threatlevel))
+		threatlevel = nearby_carbons.assess_threat(judgement_criteria, weaponcheck=CALLBACK(src, .proc/check_for_weapons))
+		//speak(nearby_carbons.real_name + text(": threat: []", threatlevel))
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/bot/ed209bot.dm:48]: Just remove this comment

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Quick last few comments

**Review by Mothblocks — APPROVED** [review]

**Pull request merged**